### PR TITLE
kokkos: new test API

### PR DESCRIPTION
**Pull request by AcriusWinter**
Update standalone testing API. 

Supersedes #35807 (for one package)

As of commit https://github.com/spack/spack/pull/45010/commits/0c0434fb908f5323af56e6eebb6c19af8b20189e:

Test output:
```
$ spack -v test run kokkos
==> Spack test bpb7pgaa4vfrqo7zcpv6yuclmp2soyfc
==> Testing package kokkos-4.0.01-3tpxp6g
...
==> [2024-07-18-20:05:36.279176] test: test_run: Test if kokkos builds and runs
...
Running tests...
Test project $HOME/.spack/test/bpb7pgaa4vfrqo7zcpv6yuclmp2soyfc/kokkos-4.0.01-3tpxp6g
    Start 1: hello_world
1/4 Test #1: hello_world ......................   Passed    0.02 sec
    Start 2: simple_reduce
2/4 Test #2: simple_reduce ....................   Passed    0.01 sec
    Start 3: random_numbers
3/4 Test #3: random_numbers ...................   Passed    0.01 sec
    Start 4: dual_view
4/4 Test #4: dual_view ........................   Passed   14.44 sec

100% tests passed, 0 tests failed out of 4

Label Time Summary:
Kokkos    =  14.47 sec*proc (4 tests)

Total Test time (real) =  14.51 sec
PASSED: Kokkos::test_run
==> [2024-07-18-20:06:06.383891] Completed testing
==> [2024-07-18-20:06:06.384033] 
======================== SUMMARY: kokkos-4.0.01-3tpxp6g ========================
Kokkos::test_run .. PASSED
============================= 1 passed of 1 parts ==============================
==> Testing package kokkos-4.0.00-bkwknh4
...
==> [2024-07-18-20:06:15.057181] test: test_run: Test if kokkos builds and runs
...
Running tests...
Test project $HOME/.spack/test/bpb7pgaa4vfrqo7zcpv6yuclmp2soyfc/kokkos-4.0.00-bkwknh4
    Start 1: hello_world
1/4 Test #1: hello_world ......................   Passed    0.02 sec
    Start 2: simple_reduce
2/4 Test #2: simple_reduce ....................   Passed    0.01 sec
    Start 3: random_numbers
3/4 Test #3: random_numbers ...................   Passed    0.01 sec
    Start 4: dual_view
4/4 Test #4: dual_view ........................   Passed   12.26 sec

100% tests passed, 0 tests failed out of 4

Label Time Summary:
Kokkos    =  12.30 sec*proc (4 tests)

Total Test time (real) =  12.32 sec
PASSED: Kokkos::test_run
==> [2024-07-18-20:06:40.687763] Completed testing
==> [2024-07-18-20:06:40.687890] 
======================== SUMMARY: kokkos-4.0.00-bkwknh4 ========================
Kokkos::test_run .. PASSED
============================= 1 passed of 1 parts ==============================
============================= 2 passed of 2 specs ==============================
```

### CODE REVIEW

**Comment by AcriusWinter** [comment]
PASSED: Kokkos::test_run
==> [2024-07-02-16:01:29.330556] Completed testing
==> [2024-07-02-16:01:29.330761] 
======================== SUMMARY: kokkos-4.2.01-ythg46p ========================
Kokkos::test_build .. PASSED
Kokkos::test_run .. PASSED
============================= 2 passed of 2 parts ==============================
============================== 1 passed of 1 spec ==============================

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 1]
```diff
@@ -408,39 +406,35 @@ def setup_build_tests(self):
...
+        exe(cmake_path, "-DEXECUTABLE_OUTPUT_PATH=" + cmake_path)
 
-    def run_tests(self, cmake_path):
-        """Run test."""
-        if not self.run_test(
-            "make", options=[cmake_path, "test"], purpose="Checking ability to execute."
-        ):
-            tty.warn("Failed to run kokkos test")
+        make = which("make")
+        make()
```
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py:409]: Are we sure that `CMake` will use `Makefile generator`?

It is true that the previous code was explicitly using `make` also, but perhaps there is a more robust way now?

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 2]
```diff
@@ -408,39 +406,35 @@ def setup_build_tests(self):
...
 
-    def build_tests(self, cmake_path):
+    def test_build(self):
         """Build test."""
-        cmake_bin = self.cmake_bin(set=False)
-
-        if not cmake_bin:
-            tty.msg("Skipping kokkos test: cmake_bin_path.txt not found")
-            return
+        # Skip if unsupported version
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/kokkos/package.py]: This comment doesn't match the implementation.

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 3]
```diff
@@ -408,39 +406,35 @@ def setup_build_tests(self):
...
         self.cmake_bin(set=True)
 
-    def build_tests(self, cmake_path):
+    def test_build(self):
         """Build test."""
-        cmake_bin = self.cmake_bin(set=False)
-
-        if not cmake_bin:
-            tty.msg("Skipping kokkos test: cmake_bin_path.txt not found")
-            return
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/kokkos/package.py:417]: This is definitely the `cmake` hack that needs to be removed -- along with the `cmake_bin` method (lines 375-387) -- but the substitution is `self.spec["cmake"].command`.  See the second note in the docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#building-and-running-test-executables.

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 4]
```diff
@@ -408,39 +406,35 @@ def setup_build_tests(self):
         self.cache_extra_test_sources(cmake_out_path)
         self.cmake_bin(set=True)
 
-    def build_tests(self, cmake_path):
+    def test_build(self):
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/kokkos/package.py]: This isn't a test part; rather it should stay `build_tests()` but with a replacement of the cmake hack..

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 1 above)
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/kokkos/package.py:409]: The options that should be passed to `make()` are `cmake_path, "test"` as shown on old line 431.

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 4 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/kokkos/package.py]: Actually, I would get rid of this method altogether, moving the call to `cmake` and `make` into the test method.  The test should be cleaner without the `cmake_bin` hack.

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 5]
```diff
@@ -408,39 +406,35 @@ def setup_build_tests(self):
...
         )
 
         if not os.path.exists(cmake_path):
-            tty.warn("Skipping smoke tests: {0} is missing".format(cmake_path))
-            return
+            raise SkipTest(f"{cmake_path} is missing")
 
-        self.build_tests(cmake_path)
-        self.run_tests(cmake_path)
+        exe = which("make")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/kokkos/package.py]: This should be building an executable for the test script.  Perhaps `spack_test_out = which(cmake_path)` (based on the value of `self.test_script_relative_path`).

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 6]
```diff
@@ -408,39 +406,35 @@ def setup_build_tests(self):
...
+        exe(cmake_path, "test")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/kokkos/package.py]: AFAICT There is no `"test"` argument at old line 446.  So this should be, assuming the suggestion for line 439, `spack-test_out(cmake_path)`.

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 7]
```diff
@@ -408,39 +406,35 @@ def setup_build_tests(self):
...
-            tty.warn("Skipping kokkos test: failed to generate Makefile")
-            return
+        cmake_bin = self.cmake_bin(set=False)
+        if not cmake_bin:
+            assert False, "cmake_bin_path.txt not found"
 
-        if not self.run_test("make", purpose="Build test software"):
-            tty.warn("Skipping kokkos test: failed to build test")
+        exe = which(cmake_bin)
+        exe(cmake_path, "-DEXECUTABLE_OUTPUT_PATH=" + cmake_path)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/kokkos/package.py]: Would be cleaner IMO if this -- as in all of the `cmake` and `make` code -- was moved to the test method and the name of the executable be `cmake` instead of `exe`.  Also, the `Executable` instance should be assigned `self.spec["cmake"].command` as calling `which(<path>)` is not needed.

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Good attempt.  The cmake hack needs to be fully removed in favor of the preferred way to access the correct `cmake`.  See other comments for more.

**Comment by AcriusWinter** [comment]
Tests 4.2.01:
```
======================== SUMMARY: kokkos-4.2.01-ythg46p ========================
Kokkos::test_run_dual_view .. PASSED
Kokkos::test_run_hello_world .. PASSED
Kokkos::test_run_simple_reduce .. PASSED
Kokkos::test_run .. PASSED
============================= 4 passed of 4 parts ==============================
============================== 1 passed of 1 spec ==============================
```

Test 3.6.01:
```
======================== SUMMARY: kokkos-3.6.01-ayaa7lz ========================
Kokkos::test_run_dual_view .. PASSED
Kokkos::test_run_hello_world .. PASSED
Kokkos::test_run_simple_reduce .. PASSED
Kokkos::test_run .. PASSED
============================= 4 passed of 4 parts ==============================
============================= 2 passed of 2 specs ==============================
```

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/kokkos/package.py:409]: > The options that should be passed to `make()` are `cmake_path, "test"` as shown on old line 431.

That is to *run* the tests -- so *after*  line 409's `make()` to build them.

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 8]
```diff
@@ -406,41 +392,26 @@ def setup_build_tests(self):
...
 
-        self.build_tests(cmake_path)
-        self.run_tests(cmake_path)
+        with working_dir(cmake_path):
+            for program in ["dual_view", "hello_world", "simple_reduce"]:
+                with test_part(self, "test_run_" + program, purpose=f"Run {program}"):
+                    exe = which(program)
+                    if exe is None:
+                        raise SkipTest(f"Test {program} not found")
+                    exe()
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/kokkos/package.py]: Why was this added?  What happens if they introduce more tests in newer versions that would be run with the original `make(cmake_path, "test")`?

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/kokkos/package.py:409]: What do you recommend?

**Review by tldahlgren — APPROVED** [review]

File: var/spack/repos/builtin/packages/kokkos/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/kokkos/package.py:409]: Closing comment.  Please feel free to submit a new PR if additional changes are needed.

**Review by tldahlgren — APPROVED** [review]
